### PR TITLE
[#1909] Add units to item weight & properly convert for calculating encumbrance

### LINKED ISSUE #1909: Add units to item weight
Currently all items are assumed to be weighed in pounds. This makes it difficult to represent large vehicle equipment that might weigh tons or to support a metric mode in the system. I propose converting weight into an object with a `units` parameter:

```javascript
weight: {
  value: 5,
  units: "lbs"
}
```

This would also require defining different weight units on config:

```javascript
DND5E.weightUnits = {
  lb: {
    label: "Pounds",
    abbreviation: "lb",
    conversion: 1
  },
  tn: {
    label: "Tons",
    abbreviation: "tn",
    conversion: 2000
  },
  kg: {
    label: "Kilograms",
    abbreviation: "kg",
    conversion: 2.2046
  },
  t: {
    label: "Tonnes",
    abbreviation: "t",
    conversion: 2204.6
  }
};
```

Calculating weight for character sheets would be a simple matter of multiplying the weight `value` by the conversion multiplier specified its unit's config.

### PR DESCRIPTION
**Pull request by arbron**
Item weights now define a specific unit which will make it possible to have a mixture of imperial and metric object weights, as well as represent large items in vehicle inventories. Defines these weights along with a conversion method in `CONFIG.weightUnits`.

Closes #1909

### To Do
- [ ] Convert SRD items

### CODE REVIEW

**Comment by p4535992** [comment]
@arbron do you think the same concept can make sense for meters and feet as well ?

**Comment by erikhuck** [comment]
fixes #1237

**Comment by erikhuck** [comment]
@arbron Does this PR enable changing the units of measurement of vehicles? I don't think having tons as the only unit of measurement really makes sense for dnd5e vehicles.

**Comment by Fyorl** [comment]
Going to try to get this and #3415 in 3.2 as I understand the encumbrance levels are causing some disruption for metric users. If I understand the gist of this PR, it's similar to how we added units to currency, but for weight now too?

The new sheets at the moment simply use an icon to represent weight, rather than a unit, and I'd prefer to keep it that way as I think it's a bit easier to parse visually. Is it reasonable to assume users won't be mixing and matching items with different weights on their sheets? Or is this likely to become a point of confusion?

**Comment by krbz999** [comment]
Mixing lb and kg would be strange to do (in my opinion). But if it ends up being allowed, a simple tooltip could be added on the weight of the item on the actor sheet.

**Comment by Fyorl** [comment]
Yeah, a tooltip would be no problem

File: lang/en.json [hunk 1]
```diff
@@ -1817,6 +1817,25 @@
...
 "DND5E.WeaponSimpleR": "Simple Ranged",
 "DND5E.Weight": "Weight",
+"DND5E.WeightUnit": {
+  "Label": "Weight Units",
+  "Kilograms": {
+    "Label": "Kilograms",
+    "Abbreviation": "kg"
+  },
+  "Megagrams": {
+    "Label": "Tonnes",
```
Comment (krbz999) [inline lang/en.json:1827]: What's wrong with "Megagram" and "mg"?

File: lang/en.json [hunk 2]
```diff
@@ -1817,6 +1817,25 @@
 "DND5E.WeaponSimpleProficiency": "Simple",
 "DND5E.WeaponSimpleR": "Simple Ranged",
 "DND5E.Weight": "Weight",
+"DND5E.WeightUnit": {
+  "Label": "Weight Units",
+  "Kilograms": {
```
Comment (krbz999) [inline lang/en.json:1822]: "Kilograms" and "Megagrams" are not actual words. The plural forms are "Kilogram" and "Megagram".

File: module/documents/actor/actor.mjs [hunk 3]
```diff
@@ -501,18 +501,20 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
   _prepareEncumbrance() {
     const config = CONFIG.DND5E.encumbrance;
     const encumbrance = this.system.attributes.encumbrance ??= {};
-    const units = game.settings.get("dnd5e", "metricWeightUnits") ? "metric" : "imperial";
+    const baseUnits = CONFIG.DND5E.encumbrance.baseUnits[this.type] ?? CONFIG.DND5E.encumbrance.baseUnits.default;
+    const unitSystem = game.settings.get("dnd5e", "metricWeightUnits") ? "metic" : "imperial";
```
Comment (krbz999) [inline module/documents/actor/actor.mjs:505]:
```diff
- const unitSystem = game.settings.get("dnd5e", "metricWeightUnits") ? "metic" : "imperial";
+ const unitSystem = game.settings.get("dnd5e", "metricWeightUnits") ? "metric" : "imperial";
```

File: module/config.mjs [hunk 4]
```diff
@@ -1805,9 +1850,15 @@ DND5E.encumbrance = {
...
-    imperial: 2000, // 2000 lbs in an imperial ton
-    metric: 1000 // 1000 kg in a metric ton
+  baseUnits: {
+    default: {
+      imperial: "lb",
+      metric: "kg"
+    },
+    vehicle: {
+      imperial: "tn",
+      metric: "mg"
```
Comment (krbz999) [inline module/config.mjs:1860]: Using "mg" here so presumably just a mistake up above.

File: lang/en.json (hunk 2 above)
Comment (thatlonelybugbear) [reply lang/en.json:1822]: > "Kilograms" and "Megagrams" are not actual words. The plural forms are "Kilogram" and "Megagram".

I don't believe that this is correct. Kilograms and Megagrams are totally valid plural forms 👍

File: lang/en.json (hunk 1 above)
Comment (arbron) [reply lang/en.json:1827]: I thought `Tonnes` was the more generally used term, but I'd welcome feedback from someone in a metric country :)

File: module/config.mjs (hunk 4 above)
Comment (arbron) [reply module/config.mjs:1860]: What mistake are you referring to?

Comment (krbz999) [reply module/config.mjs:1860]: Using "t" as the abbreviation for Megagrams in the localization.

Comment (arbron) [reply module/config.mjs:1860]: This is the static `key`, not the localization value, so it references the key in `CONFIG.weightUnits`.

File: lang/en.json (hunk 1 above)
Comment (krbz999) [reply lang/en.json:1827]: I am, lol, but here's another to make a fool of me.

**Comment by GamerFlix** [comment]
Just to comment on the megagrams point and suggested alternative:

According to convention (compare: https://en.wikipedia.org/wiki/Metric_prefix) `mg` would stand for miligrams while megagrams (which I have never heard uses before) would be `Mg` (note the case). I would also like to note that tons vs tonnes for metric vs imperial seems rather confusing to me when it could simply be Imperial tons vs Metric tons for perfect clarity. The latter point is especially noteworthy for people for whom English is not their first language and therefore aren't familiar with the spelling difference's meaning.

**Review by Fyorl — APPROVED** [review]
Looks good, thanks. Please see my comments but let's get this merged and get some testing in.

**Pull request merged**